File: core/workers/audiodb.py
"""
audiodb.py — TheAudioDB enrichment worker.

Gratis publieke API, geen auth nodig.
Rate limit: conservatief 1 call per 3 sec.
"""

import logging
import time
from typing import Any, Optional

import requests

from core.workers.base import BaseWorker

AUDIODB_BASE    = "https://theaudiodb.com/api/v1/json/2"
RATE_INTERVAL   = 3.0
REQUEST_TIMEOUT = 12


class AudioDBWorker(BaseWorker):
    source = "audiodb"

    def __init__(self, database, logger: logging.Logger):
        super().__init__(database, logger)
        self._last_call = 0.0
        self._session   = requests.Session()
        self._session.headers.update({
            "Accept":     "application/json",
            "User-Agent": "LastfmMuziekApp/1.0",
        })

    def _rate_limit(self) -> None:
        wait = max(0.0, RATE_INTERVAL - (time.time() - self._last_call))
        if wait > 0:
            time.sleep(wait)
        self._last_call = time.time()

    def _get(self, url: str) -> dict:
        self._rate_limit()
        resp = self._session.get(url, timeout=REQUEST_TIMEOUT)
        resp.raise_for_status()
        return resp.json()

    def process(self, item: dict) -> dict[str, Any]:
        try:
            data = self._search(item["entity_name"], item["entity_type"])
            if not data:
                return {"ok": False, "error": "No results found"}
            return {"ok": True, "data": data}
        except Exception as exc:
            self.log.warning(f"AudioDB worker mislukt voor '{item.get('entity_name')}': {exc}")
            return {"ok": False, "error": str(exc)}

    def _search(self, name: str, entity_type: str = "artist") -> Optional[dict]:
        if entity_type == "artist":
            return self._search_artist(name)
        if entity_type == "album":
            return self._search_album(name)
        return None

    def _search_artist(self, name: str) -> Optional[dict]:
        from urllib.parse import quote
        data    = self._get(f"{AUDIODB_BASE}/search.php?s={quote(name)}")
        artists = data.get("artists") or []
        if not artists:
            return None

        norm  = lambda s: (s or "").lower().strip()
        exact = next((a for a in artists if norm(a.get("strArtist", "")) == norm(name)), None)
        a     = exact or artists[0]

        return {
            "audiodbId":     a.get("idArtist"),
            "name":          a.get("strArtist"),
            "formedYear":    a.get("intFormedYear"),
            "bornYear":      a.get("intBornYear"),
            "disbandedYear": a.get("intDisbandYear"),
            "gender":        a.get("strGender"),
            "members":       a.get("intMembers"),
            "country":       a.get("strCountry"),
            "countryCode":   a.get("strCountryCode"),
            "genre":         a.get("strGenre"),
            "style":         a.get("strStyle"),
            "mood":          a.get("strMood"),
            "theme":         a.get("strTheme"),
            "website":       a.get("strWebsite"),
            "facebook":      a.get("strFacebook"),
            "twitter":       a.get("strTwitter"),
            "biography": {
                "en": a.get("strBiographyEN"),
                "nl": a.get("strBiographyNL"),
                "de": a.get("strBiographyDE"),
                "fr": a.get("strBiographyFR"),
            },
            "logo":          a.get("strArtistLogo"),
            "thumb":         a.get("strArtistThumb"),
            "banner":        a.get("strArtistBanner"),
            "fanart":        a.get("strArtistFanart") or a.get("strArtistFanart2"),
            "musicbrainzId": a.get("strMusicBrainzID"),
            "lastfmChart":   a.get("strLastFMChart"),
            "source":        "audiodb",
            "fetchedAt":     int(time.time() * 1000),
        }

    def _search_album(self, name: str) -> Optional[dict]:
        from urllib.parse import quote
        data   = self._get(f"{AUDIODB_BASE}/searchalbum.php?s=all&a={quote(name)}")
        albums = data.get("album") or []
        if not albums:
            return None

        norm  = lambda s: (s or "").lower().strip()
        exact = next((a for a in albums if norm(a.get("strAlbum", "")) == norm(name)), None)
        al    = exact or albums[0]

        return {
            "audiodbId":     al.get("idAlbum"),
            "name":          al.get("strAlbum"),
            "artistName":    al.get("strArtist"),
            "year":          al.get("intYearReleased"),
            "genre":         al.get("strGenre"),
            "style":         al.get("strStyle"),
            "mood":          al.get("strMood"),
            "theme":         al.get("strTheme"),
            "speed":         al.get("strSpeed"),
            "description": {
                "en": al.get("strDescriptionEN"),
                "nl": al.get("strDescriptionNL"),
            },
            "thumbUrl":      al.get("strAlbumThumb"),
            "thumbBackUrl":  al.get("strAlbumThumbBack"),
            "cdArtUrl":      al.get("strAlbumCDart"),
            "spine":         al.get("strAlbumSpine"),
            "label":         al.get("strLabel"),
            "score":         al.get("intScore"),
            "loved":         al.get("intLoved"),
            "musicbrainzId": al.get("strMusicBrainzID"),
            "source":        "audiodb",
            "fetchedAt":     int(time.time() * 1000),
        }

File: .claude/worktrees/vigilant-bose-84213c/services/enrichment/workers/audiodb.js
'use strict';
// ── AudioDB worker ────────────────────────────────────────────────────────────
// API: https://theaudiodb.com/api/v1/json/2/search.php?s=ARTIST
// Gratis tier — geen API key nodig voor publieke endpoints
// Rate limit: conservatief 1 call per 3 sec

const AUDIODB_BASE  = 'https://theaudiodb.com/api/v1/json/2';
const RATE_INTERVAL = 3000;

class AudioDBWorker {
  constructor(db, log) {
    this.db  = db;
    this.log = log.child ? log.child({ worker: 'audiodb' }) : log;
    this._lastCall = 0;
  }

  async _rateLimit() {
    const wait = Math.max(0, RATE_INTERVAL - (Date.now() - this._lastCall));
    if (wait > 0) await new Promise(r => setTimeout(r, wait));
    this._lastCall = Date.now();
  }

  async _get(url) {
    await this._rateLimit();
    const res = await fetch(url, {
      headers: { 'Accept': 'application/json', 'User-Agent': 'LastfmMuziekApp/1.0' },
      signal:  AbortSignal.timeout(12_000),
    });
    if (!res.ok) throw new Error(`AudioDB HTTP ${res.status}`);
    return res.json();
  }

  async process(entity) {
    try {
      const data = await this.search(entity.entity_name, entity.entity_type);
      if (!data) return { ok: false, error: 'No results found' };
      return { ok: true, data };
    } catch (err) {
      this.log.warn({ err: err.message, entity: entity.entity_name }, 'AudioDB fetch failed');
      return { ok: false, error: err.message };
    }
  }

  async search(name, entityType = 'artist') {
    if (entityType === 'artist') {
      return this._searchArtist(name);
    } else if (entityType === 'album') {
      return this._searchAlbum(name);
    }
    return null;
  }

  async _searchArtist(name) {
    const encoded = encodeURIComponent(name);
    const data    = await this._get(`${AUDIODB_BASE}/search.php?s=${encoded}`);
    const artists = data.artists || [];
    if (!artists.length) return null;

    const norm  = s => (s || '').toLowerCase().trim();
    const exact = artists.find(a => norm(a.strArtist) === norm(name));
    const a     = exact || artists[0];

    return {
      audiodbId:     a.idArtist,
      name:          a.strArtist,
      formedYear:    a.intFormedYear || null,
      bornYear:      a.intBornYear || null,
      disbandedYear: a.intDisbandYear || null,
      gender:        a.strGender || null,
      members:       a.intMembers || null,
      country:       a.strCountry || null,
      countryCode:   a.strCountryCode || null,
      genre:         a.strGenre || null,
      style:         a.strStyle || null,
      mood:          a.strMood || null,
      theme:         a.strTheme || null,
      website:       a.strWebsite || null,
      facebook:      a.strFacebook || null,
      twitter:       a.strTwitter || null,
      biography: {
        en: a.strBiographyEN || null,
        nl: a.strBiographyNL || null,
        de: a.strBiographyDE || null,
        fr: a.strBiographyFR || null,
      },
      logo:          a.strArtistLogo || null,
      thumb:         a.strArtistThumb || null,
      banner:        a.strArtistBanner || null,
      fanart:        a.strArtistFanart || a.strArtistFanart2 || null,
      musicbrainzId: a.strMusicBrainzID || null,
      lastfmChart:   a.strLastFMChart || null,
      source:        'audiodb',
      fetchedAt:     Date.now(),
    };
  }

  async _searchAlbum(name) {
    const encoded = encodeURIComponent(name);
    const data    = await this._get(`${AUDIODB_BASE}/searchalbum.php?s=all&a=${encoded}`);
    const albums  = data.album || [];
    if (!albums.length) return null;

    const norm  = s => (s || '').toLowerCase().trim();
    const exact = albums.find(a => norm(a.strAlbum) === norm(name));
    const al    = exact || albums[0];

    return {
      audiodbId:     al.idAlbum,
      name:          al.strAlbum,
      artistName:    al.strArtist,
      year:          al.intYearReleased || null,
      genre:         al.strGenre || null,
      style:         al.strStyle || null,
      mood:          al.strMood || null,
      theme:         al.strTheme || null,
      speed:         al.strSpeed || null,
      description: {
        en: al.strDescriptionEN || null,
        nl: al.strDescriptionNL || null,
      },
      thumbUrl:      al.strAlbumThumb || null,
      thumbBackUrl:  al.strAlbumThumbBack || null,
      cdArtUrl:      al.strAlbumCDart || null,
      spine:         al.strAlbumSpine || null,
      label:         al.strLabel || null,
      score:         al.intScore || null,
      loved:         al.intLoved || null,
      musicbrainzId: al.strMusicBrainzID || null,
      source:        'audiodb',
      fetchedAt:     Date.now(),
    };
  }
}

module.exports = { AudioDBWorker };
